src/core/memory/memory.cc

```diff
@@ -20,9 +20,123 @@
 #include "singa/utils/logging.h"
 #include "singa/proto/core.pb.h"
 #include <iostream>
+/*
+int get_pos(size_t size) {
+	int result = 0;
+	while(size > 1) {
+		result++;
+		size = size/2;
+	}
+	return result;
+}
+*/
+namespace singa {
+
+CppMemPool::CppMemPool()	{
+	memPoolSize = 0;
+	freeSize = 0;
+	ppAllocUints = (struct _Uint**)malloc(64*sizeof(struct _Uint*));
+	ppFreeUints = (struct _Uint**)malloc(64*sizeof(struct _Uint*));
+	for(int i = 0; i < 64; i++) {
+		ppAllocUints[i] = NULL;
+		ppFreeUints[i] = NULL;
+	}
+}
+
+
+Block* CppMemPool::Malloc(const size_t size) {	
+	CHECK(size > 0);
+	Block *pAllocBlk = NULL;
+	int pos = 63 - __builtin_clzll(size);
+	
+	struct _Uint*& pAllocUint = ppAllocUints[pos];
+	struct _Uint*& pFreeUint = ppFreeUints[pos];
+	struct _Uint* pCurUint = NULL;
+	size_t memSize = pow(2,pos);
+	size_t blkSize = (size % memSize == 0) ? memSize : memSize*2;
+	blkSize += sizeof(struct _Uint);
+	
+	if(pFreeUint == NULL) { // if no available free blocks
+		memPoolSize += blkSize;
+		pCurUint = (struct _Uint*)malloc(blkSize);
+		pCurUint->pPrev = NULL;
+		pCurUint->pNext = pAllocUint; 
+		if(pAllocUint != NULL) {
+			pAllocUint->pPrev = pCurUint;
+		}
+		pAllocUint = pCurUint;
+		pAllocBlk = new Block((char*)(pCurUint) + sizeof(struct _Uint), size);
```

Comment:
If the requested size is 1024, then we allocated 1024+16, 16 is used for metadata of the mempool.